File: TestObjects.VB/VBCompiler/RefOut.vb
'Imports ExpressionTreeTestObjects.Dummy

'Partial Module VBCompiler
'    <TestObject(RefOut)>
'    Friend PassRef As Expression = Expr(Sub(i As Integer) DummyMethodWithRef(i))

'    <TestObject(RefOut)>
'    Friend PassOut As Expression = Expr(Sub(i As Integer) DummyMethodWithOut(i))

'    <TestObject(RefOut)>
'    Friend PassRefField = Expr(Sub(d As Dummy2) DummyMethodWithRef(d.Data))

'    <TestObject(RefOut)>
'    Friend PassOutField = Expr(Sub(d As Dummy2) DummyMethodWithOut(d.Data))

'    <TestObject(RefOut)>
'    Friend PassRefClosureVariable = IIFE(Function()
'                                             Dim i As Integer = 0
'                                             Return Expr(Sub() DummyMethodWithRef(i))
'                                         End Function)

'    <TestObject(RefOut)>
'    Friend PassOutClosureVariable = IIFE(Function()
'                                             Dim i As Integer = 0
'                                             Return Expr(Sub() DummyMethodWithOut(i))
'                                         End Function)
'End Module

File: TestObjects/FactoryMethods/RefOut.cs
//using System.Linq.Expressions;
//using static ExpressionTreeTestObjects.Categories;
//using static System.Linq.Expressions.Expression;
//using static ExpressionTreeTestObjects.Functions;

namespace ExpressionTreeTestObjects {
    //internal delegate int DelegateWithOut(out int i);
    //internal delegate int DelegateWithRef(ref int i);

    //partial class FactoryMethods {
    //    [TestObject(RefOut)]
    //    internal static readonly Expression OutParameter = IIFE(() => {
    //        var prm = Parameter(typeof(int).MakeByRefType());
    //        return Lambda(
    //            typeof(DelegateWithOut),
    //            prm,
    //            prm
    //        );
    //    });

    //    [TestObject(RefOut)]
    //    internal static readonly Expression RefParameter = IIFE(() => {
    //        var prm = Parameter(typeof(int).MakeByRefType());
    //        return Lambda(
    //            typeof(DelegateWithRef),
    //            prm,
    //            prm
    //        );
    //    });

    //    [TestObject(RefOut)]
    //    internal static readonly Expression PassRef = 
    //        Call(
    //            typeof(Dummy).GetMethod("DummyMethodWithRef"),
    //            Parameter(typeof(int))
    //        );

    //    [TestObject(RefOut)]
    //    internal static readonly Expression PassOut =
    //        Call(
    //            typeof(Dummy).GetMethod("DummyMethodWithOut"),
    //            Parameter(typeof(int))
    //        );

    //    [TestObject(RefOut)]
    //    internal static readonly Expression PassRefField =
    //        Call(
    //            typeof(Dummy).GetMethod("DummyMethodWithRef"),
    //            PropertyOrField(
    //                Constant(new Dummy2()),
    //                "Data"
    //            )
    //        );

    //    [TestObject(RefOut)]
    //    internal static readonly Expression PassOutField =
    //        Call(
    //            typeof(Dummy).GetMethod("DummyMethodWithOut"),
    //            PropertyOrField(
    //                Constant(new Dummy2()),
    //                "Data"
    //            )
    //        );
    //}
}
